### ISSUE TITLE
Fake Real Threads??

### ISSUE THREAD

**Issue by jeffRTC** [emscripten-core/emscripten#12920]
TL;DR

I have a situation where a specific API is not available within the webWorkers, and I need to have concurrency. The operations done within this API are just I/O ones. I use em_val to call this API. 

-----

Looking at history, AmigaDOS did multitasking very well without any concept of processes at all -- all programs ran in a single address space. 

More recently, Intel has introduced hyper threading which fakes threads at hardware layer.

I however failed to find any ready-made software layer solutions. 

Upon Googling more about this problem, I have collected following information, 

_**"To make a single core able to run multiple threads, a form of time-division multiplexing was used. To simplify things a bit: the OS sets up a timer which interrupts the system at a fixed interval. A single interval is known as a time slice. Everytime this interrupt occurs, the OS runs the scheduling routine, which picks the next thread that is due to be executed. The context of the core is then switched from the currently running thread to the new thread, and execution continues."**_

_**"Instruction-level parallelism allows a sequence of instructions derived from a sequential program to be parallelized for execution on multiple pipelined functional units. If industry acceptance is a measure of importance, ILP has blossomed. It now profoundly influences the design of almost all leading-edge microprocessors and their compilers. ILP's key advantage is that it exploits parallelism without requiring the programmer to rewrite existing applications. This is attractive because today's applications are still programmed sequentially, and many will never be rewritten. Exploiting ILP across a diverse set of performance-critical applications will require renewed emphasis on the role of the compiler. To advance, ILP compilers will require an enormous research effort, much like the one that drove vector compilers to today's relatively mature status. Thus far, however, ILP has yet to receive a similar investment, even though it presents what may be even more complex technical challenges. A number of ILP compilers exist both in academia and in industry, yet the development of ILP is far from complete."**_

Further exploring this problem within Emscripten own Github,  I have collected following comments,

@TerrorJack wrote,

**_To achieve this, we can implement a binaryen pass much like the existing asyncify pass to rewrite all blocking call sites to save context and yield execution. (Or possibly at the clang level) A minimal round-robin scheduler can be implemented in C to handle the thread switching logic._**

@sbc100 wrote,

**_What you are describing is more like coroutines or fibres. You can indeed achieve something like that using asyncify, but it not really a true pthread replacement as there is no pre-preemption._**

**_I remember a project from a while back that tried to do this (fake real threads using coroutines): https://www.gnu.org/software/pth/. But its not a completely drop in replacement for real pre-preemptive threads and I think that project is inactive these days. I think its based on setcontext/getcontext so presumably one starting point would be take that GNU pth and use asyncify to implement setcontext/getcontext._**

------------

I'm looking somewhat automatically managed concurrent instructions execution method.

Looking at Fibers and  Coroutines, It appear that I have manage everything manually my own.

However, Coroutines looks promising.

According to a one JavaScript library [http://js-coroutines.com/](url) it claims that it has **power to sort, stringify, compress, decompress and run any custom complex methods while staying smooth and interactive at 60fps**.

**Comment by jeffRTC** [comment]
@kripken  On Twitter, you wrote that _**"We should probably add an Asyncify "backend" option that uses atomics and has almost no overhead. The reason we haven't focused on that is just because atomics aren't supported everywhere yet."**_

I think this is a game changer. Is there any pull requests with this feature I could merge?

**Comment by sbc100** [comment]
We already have a fiber API build in emscripten which you should be able to use a starting point for a scheduler where you can manage your "somewhat automatically managed concurrent instruction".     See https://emscripten.org/docs/api_reference/fiber.h.html.

You would represent each of your "processes" as a fiber and switch between them whenever you choose.  The one thing you can't do is interrupt long running process that don't yield in any way or make any calls.  This was is also true of AmigaDOS of course, it requires some cooperation from the "processes" .. i.e. they are not preempt-able.

**Comment by jeffRTC** [comment]
@sbc100 

I aware of Fiber API but it requires Asyncify (Which has overhead) 

But if you or someone add Asyncify backend option that use atomics, it would remove this cost.

**Comment by sbc100** [comment]
> @sbc100
> 
> I aware of Fiber API but it requires Asyncify (Which has overhead)
> 
> But if you or someone add Asyncify backend option that use atomics, it would remove this cost.

atomics only make sense in contents of threads or workers.  If your environment doesn't have workers can't take advantage of atomics.

**Comment by jeffRTC** [comment]
But I see atomics are available on main thread in Chrome?!?

**Comment by sbc100** [comment]
Yes, you can use atomics on the main thread, but in single threaded programs they have no effect.   They are only meaningful when communicating with other threads.

**Comment by jeffRTC** [comment]
`Atomics.waitAsync, the async version of Atomics.wait. Atomics.waitAsync allows programmers to wait on a SharedArrayBuffer location in the same fashion as Atomics.wait but returns a Promise instead.

Would waitAsync has any effect? 


If no, is it impossible to reduce the overhead of asyncify on single thread?
`

**Comment by sbc100** [comment]
Yes, that is correct.    Most of the ways to improve asyncify at this point involve new proposals to the wasm spec that can be used to implement some kind of stack switching/suspension.   As far as I know there is nothing better (than asyncify) we can do with the tools we have the browsers today.  

IIUC I can't how `Atomics.waitAsync` can help in the single threaded mode either.

**Comment by jeffRTC** [comment]
Does em:val works with fibers?

**Comment by jeffRTC** [comment]
@sbc100 

The fiber.h API quite low level to me. Is it possible to use Boost fiber or other high level abstractions? 

I looked into Boost and they seems using their own context APis

**Comment by sbc100** [comment]
It depends what boost use for their fibers.   Most implementations of co-routines/fibers tend to use either assembly, or setcontext/getcontext, neither of which would work with emscripten.    I might just work though, I have no idea what then depend on.

**Comment by jeffRTC** [comment]
They are using ASM :(

https://github.com/boostorg/context/tree/develop/src/asm
